hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/protocol/StorageContainerLocationProtocolServerSideTranslatorPB.java

```diff
@@ -172,9 +191,185 @@ public ScmContainerLocationResponse submitRequest(RpcController controller,
+      }
+    }
+    ScmContainerLocationResponse response = dispatcher
         .processRequest(request, this::processRequest, request.getCmdType(),
             request.getTraceID());
+    if (checkResponseForECRepConfig) {
+      switch (response.getCmdType()) {
+      case GetContainer:
+        disallowECReplicationConfigInGetContainerResponse(response);
+        break;
+      case ListContainer:
+        disallowECReplicationConfigInListContainerResponse(response);
+        break;
+      case GetContainerWithPipeline:
+        disallowECReplicationConfigInGetContainerWithPipelineResponse(response);
+        break;
+      case GetContainerWithPipelineBatch:
+        disallowECReplicationConfigInGetContainerWithPipelineBatchResponse(
+            response);
+        break;
+      case GetExistContainerWithPipelinesInBatch:
+        disallowECReplicationConfigInGetExistContainerWithPipelineBatchResponse(
+            response);
+        break;
+      case ListPipelines:
+        disallowECReplicationConfigInListPipelinesResponse(response);
+        break;
+      case GetPipeline:
+        disallowECReplicationConfigInGetPipelineResponse(response);
+        break;
+      default:
+      }
+    }
+    return response;
+  }
+
+  private void disallowECReplicationConfigInListContainerResponse(
+      ScmContainerLocationResponse response) throws ServiceException {
+    if (!response.hasScmListContainerResponse()) {
+      return;
+    }
+    for (HddsProtos.ContainerInfoProto containerInfo :
+        response.getScmListContainerResponse().getContainersList()) {
+      if (containerInfo.hasEcReplicationConfig()) {
+        throw new ServiceException(ERROR_LIST_CONTAINS_EC_REPLICATION_CONFIG);
+      }
+    }
+  }
+
+  private void disallowECReplicationConfigInGetContainerResponse(
+      ScmContainerLocationResponse response) throws ServiceException {
+    if (!response.hasGetContainerResponse()) {
+      return;
+    }
+    if (!response.getGetContainerResponse().hasContainerInfo()) {
+      return;
+    }
+    if (response.getGetContainerResponse().getContainerInfo()
+        .hasEcReplicationConfig()) {
+      throw new ServiceException(ERROR_RESPONSE_CONTAINS_EC_REPLICATION_CONFIG);
+    }
+  }
+
+  private void disallowECReplicationConfigInGetContainerWithPipelineResponse(
+      ScmContainerLocationResponse response) throws ServiceException {
+    if (!response.hasGetContainerWithPipelineResponse()) {
+      return;
+    }
+    if (!response.getGetContainerWithPipelineResponse()
+        .hasContainerWithPipeline()) {
+      return;
+    }
+    if (response.getGetContainerWithPipelineResponse()
+        .getContainerWithPipeline().hasContainerInfo()) {
+      HddsProtos.ContainerInfoProto containerInfo =
+          response.getGetContainerWithPipelineResponse()
+              .getContainerWithPipeline().getContainerInfo();
+      if (containerInfo.hasEcReplicationConfig()) {
+        throw new ServiceException(
```

Comment:
Discussed offline, but we should check if the admin CLI will do retries for this (which we don't want) because the type is ServiceException.

Reply:
I did a quick test today, and found that if we throw a plain ServiceException from the submitRequest method here, we run into an NP in the hadoop proto layer that the SCM admin interface uses as it seems...
This is the exception on the SCM side:
`2022-04-19 09:58:14,181 [IPC Server handler 22 on default port 9860] WARN ipc.Server: IPC Server handler 22 on default port 9860, call Call#0 Retry#0 org.apache.hadoop.hdds.scm.protocol.StorageContainerLocationProtocol.submitRequest from 172.20.0.8:42982
java.lang.NullPointerException
        at org.apache.hadoop.ipc.ProtobufRpcEngine$Server.processCall(ProtobufRpcEngine.java:477)
        at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:574)
        at org.apache.hadoop.ipc.ProtobufRpcEngine2$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine2.java:552)
        at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:1093)
        at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:1035)
        at org.apache.hadoop.ipc.Server$RpcCall.run(Server.java:963)
        at java.base/java.security.AccessController.doPrivileged(Native Method)
        at java.base/javax.security.auth.Subject.doAs(Subject.java:423)
        at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1878)
        at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2966)
`
This is because the hadoop layer in ProtobufRpcEngine$Server.processCall checks the ServiceException cause and would like to add a metric for that, but the ServiceException we throw does not have a cause...

I have modified this code to throw an SCMException with INTERNAL_ERROR as the error code, and I am adding an UNSUPPORTED_OPERATION as a resultcode here, as we do not have one.
Tested the changes manually, and now on the client side I see the INTERNAL_ERROR reported with the proper error message instead of the NPE.